Task/zen 2464 - Documentation for SCG GenTemplates (DO NOT MERGE)

andylowry opened the pull request:
@tedepstein I've now created documentation on using SCG Gentemplates in general, and on creating custom SCG GenTemplates. Please include in your review.

This PR should not be merged until the other two PRs have been merged, since it builds on their commits.

andylowry committed:
ZEN-2464 Cleaned up project a bit

Fixed file modes on scripts, and got rid of feed.xml since we're not
using it (maybe in the future), and it changes on every build due to a
timestamp it contains.

andylowry committed:
ZEN-2464 Document how to use SCG Gentemplates

Also introduced a new {%link%} liquid tag, which emits a properly
formatted link to a portal page given its slug or permalink value.

andylowry committed:
ZEN-2464 Completed Using SCG Modules article

andylowry committed:
ZEN-2464 Completed custom SCG module doc

andylowry committed:
ZEN-2464 Proofread the custom SCG article

andylowry committed:
ZEN_2471 Address comments from Ted

tedepstein commented on codegen/csg.adoc:
```diff
@@ -8,4 +8,193 @@ permalink: /codegen_custom_swagger_gentemplate/
 :RAS: RepreZen API Studio 
```
@andylowry , the name of this file is `csg.adoc`.  Was it supposed to be `csg.adoc` ?

tedepstein replied:
Or maybe `custom-scg.adoc`.  Doesn't matter that much, just helps to be able to look at the filename and know what it is.

tedepstein commented on codegen/swagger-codegen.adoc:
```diff
@@ -0,0 +1,65 @@
+service scaffolding, also geared toward a variety of languages and underlying frameworks. You'll be
+glad to know that all of these standard SCG modules are available as GenTemplates in {RAS}.
+
+== Model Type Determines GenTemplates
```
@andylowry , just a heads-up that we will want to change this, making SCG templates available for RAPID-ML files.  This will use a RAPID-->Swagger GenTarget as a prerequisite.  You don't need to make note of that in this doc right now, but we'll want to address this fairly soon.  Some of it depends on whether our current or future adopters (hopefully near-future, through SBC Fintech) want to use RAPID-ML.  RAPID-ML looks a lot more viable if it has pseudo-direct access to all of the SCG modules.

tedepstein committed:
Minor edits to 'Using Swagger-Codegen' doc

 - Rephrased the Swagger-OpenAPI intro to make it read a little more smoothly.
 - Corrected JSON --> YAML

tedepstein committed:
Minor edits to Custom SCG Module doc

In particular, we should resist the temptation to point out a shortcoming or inconveniences, and promise a future fix.  It makes the product appear incomplete, and puts pressure on us to follow through with promised improvements, which may compete with more important things.

tedepstein commented on codegen/csg.adoc:
```diff
@@ -8,4 +8,193 @@ permalink: /codegen_custom_swagger_gentemplate/
+
+[source%nowrap]
+--
+java -jar ~/git/swagger-codegen/modules/swagger-codegen-cli/target/swagger-codegen-cli.jar meta -o ~/LexiconSCGentemplates -n lexiconSCGenTemplate -p my.org
```
We haven't pointed out yet that we want to put our SCG meta output into a project directory that is somewhere within the Eclipse/RepreZen workspace.  Usually directly off the workspace root, as it's done here.  

But this also assumes that we've built our swagger-codegen-cli directly in the workspace, and not done anything (e.g. with classpath, maybe) to make it more easily accessible.  This makes the command line longer and harder to read.  

Maybe it would be better to add a step that makes the cli jar easier to reference; and also point out, maybe in the `-o` option above, that the output should go into a workspace subdirectory.

andylowry replied:
@tedepstein Sorry to say that almost everything you said here is not true! :)

First, the project directory need not be - and is not in the steps I've outlined - within the workspace. Eclipse (and therefore API Studio) supports "linked" projects that live outside the workspace root. In fact, these days there are probably more linked projects in the world than otherwise, since so many projects are imported from git clones that are created outside the workspace.

It's also not true that the scg-cli project lives - or needs to live - within the workspace, or that there are any classpath issues that arise with it. The jar file that's produced in that project is an "executable" jar file, which means it defines its own class path as well as an "entry point" class. When you run `java -jar xxx.jar`, Java loads that jar file, sets the classpath (ignoring any other ways of specifying classpath like environment variable or `-cp` options), and attempts to run the entry point class. So it doesn't matter where this jar file ends up, you just have to identify it correctly with the `-jar` option for it to work.

The `-o` option here really is a mouthful, but it also represents the most likely location for anyone following along with this. I expect that anyone attempting to create their own custom SCG module will completely understand what's going on here and simplify if they care to.

The real culprit here, IMO, is the swagger-codegen folks, who ship SCG with a couple dozen shell scripts that each make it easy to run a specific module against the petstore example and stick it in a specific output folder - all hard-wired. Yet they couldn't be bothered to provide a very simple but more general script that hides that big, ugly jar file path and leaves all the options open to the user!

tedepstein replied:
OK, in that case you can leave it as-is.

tedepstein commented on codegen/csg.adoc:
```diff
@@ -8,4 +8,193 @@ permalink: /codegen_custom_swagger_gentemplate/
+
+image::import-maven-step-1.png[Import Skeleton Project Step 1,role=text-center]
+
+Browse to the output folder you specified for the SCG `meta` command, and you should see the Maven
```
I think this step assumes you've placed the output somewhere within the workspace.  It wouldn't hurt to reiterate that point here.  What happens if you try to import something that's outside the workspace?  Is there an additional step to link it in...?

andylowry replied:
No, the project import will by default create a linked project when the source folder is not in the workspace. I'm pretty sure there's also an option copy everything into the workspace.

tedepstein commented:
@andylowry , my review is complete.  I've made some edits and given a few comments.  Please address these when you can.

This is very good, BTW.  Really nice docs for developers who want to do custom code gen, or understand how to work with the Swagger GenTarget configs.

tedepstein committed:
More minor edits

andylowry commented on codegen/swagger-codegen.adoc:
```diff
@@ -32,7 +32,7 @@ RAPID-ML GenTemplates
-The GenTarget (`.gen`) files that define GenTargets in a model project are JSON files that, among
+The GenTarget (`.gen`) files that define GenTargets in a model project are YAML files that, among
 other things, provide values for parameters that may be used by the GenTemplate. In general, a
 GenTemplate declares its parameters including the following information for each parameter:
```
Oops - thanks!

andylowry commented:
@ted Thanks for the edits. Should I merge?

andylowry committed:
Merge pull request #5 from RepreZen/task/ZEN-2471c

Task/zen 2471c Address comments from Ted (DO NOT MIGRATE)

andylowry committed:
Merge remote-tracking branch 'origin/master' into task/ZEN-2464
